zkevm-circuits/src/taiko_pi_circuit.rs

```diff
@@ -229,83 +259,99 @@ impl<F: Field> SubCircuitConfig<F> for TaikoPiCircuitConfig<F> {
+        cb.preload_tables(meta,
+            &[
+               (PiCellType::Lookup(Keccak), &keccak_table), 
+               (PiCellType::Lookup(Bytecode), &byte_table), 
+               (PiCellType::Lookup(Block), &block_table)
+               ]
+           );
+
         // field bytes
         meta.create_gate(
-            "rpi_field_bytes_acc[i+1] = rpi_field_bytes_acc[i] * t + rpi_bytes[i+1]",
+            "PI acc constraints", 
             |meta| {
-                let mut cb = BaseConstraintBuilder::new(MAX_DEGREE);
-
-                let q_field_step = meta.query_selector(q_field_step);
-                let rpi_field_bytes_acc_next =
-                    meta.query_advice(rpi_field_bytes_acc, Rotation::next());
-                let rpi_field_bytes_acc = meta.query_advice(rpi_field_bytes_acc, Rotation::cur());
-                let rpi_field_bytes_next = meta.query_advice(rpi_field_bytes, Rotation::next());
-                let is_field_rlc = meta.query_fixed(is_field_rlc, Rotation::next());
-                let randomness = challenges.evm_word();
-                let t = select::expr(is_field_rlc, randomness, BYTE_POW_BASE.expr());
-                cb.require_equal(
-                    "rpi_field_bytes_acc[i+1] = rpi_field_bytes_acc[i] * t + rpi_bytes[i+1]",
-                    rpi_field_bytes_acc_next,
-                    rpi_field_bytes_acc * t + rpi_field_bytes_next,
-                );
-                cb.gate(q_field_step)
-            },
-        );
-        meta.create_gate("rpi_field_bytes_acc[0] = rpi_field_bytes[0]", |meta| {
-            let mut cb = BaseConstraintBuilder::new(MAX_DEGREE);
-
-            let q_field_start = meta.query_selector(q_field_start);
-            let rpi_field_bytes_acc = meta.query_advice(rpi_field_bytes_acc, Rotation::cur());
-            let rpi_field_bytes = meta.query_advice(rpi_field_bytes, Rotation::cur());
-
-            cb.require_equal(
-                "rpi_field_bytes_acc[0] = rpi_field_bytes[0]",
-                rpi_field_bytes_acc,
-                rpi_field_bytes,
-            );
-            cb.gate(q_field_start)
+            circuit!([meta, cb], {
+                // rpi_field_bytes_acc[i+1] = rpi_field_bytes_acc[i] * t + rpi_bytes[i+1]
+                ifx!(q!(q_field_step) => {
+                    let t = ifx!(f!(is_field_rlc) => {
+                        challenges.evm_word()
+                    } elsex {
+                        BYTE_POW_BASE.expr()
+                    });
+                    require!(
+                        a!(rpi_field_bytes_acc, 1) => a!(rpi_field_bytes_acc) * t + a!(rpi_field_bytes, 1)
+                    );
+                });
+                // rpi_field_bytes_acc[0] = rpi_field_bytes[0]
+                ifx!(q!(q_field_start) => {
+                    require!(a!(rpi_field_bytes_acc) => a!(rpi_field_bytes));
+                });
+            });
+            cb.build_constraints(None)
```

Comment:
Why did you add an option for an additional selector here? People should just use `ifx` instead no?

Reply:
Because I don't like too much indentation with lots of `ifx` 😅 (aesthetically unpleasing given macro's ! and parentheses) and I feel like if there's must be a fixed selector enabling of circuits, why don't we generate the constraint "anchoring" this gate.

Reply:
If you write normal rust code, would you introduce a gate like construct to make it cleaner? If the answer is no, why would you do it here? Does normal rust code have too many nested if's? I don't think so, you just split things up over multiple functions in a way that makes sense. So why not do exactly the same with circuits? The whole of MPT is written using a single gate, I don't think there was any problem there with this because it's split up mostly like you would split up a normal rust program.

>  why don't we generate the constraint "anchoring" this gate.

Because then the flow control is split into multiple parts and gates don't allow if/else statements, so I find it harder to read and it's not something normal programmers are used to.

It's also not possible to do gates inside gates. So even if you like using gates, it's impossible to use it except on the top level of the program. So it's a strange concept to rely on because if you like to use it to split up your program, it only works for tiny programs that have only 1 level complexity, anything more complex does require using if/else statement.

It's also not directly compatible with regions/context management, so this would require a different path to make it behave like the normal control flow using ifs/matches.